Replace auth.txt with SQLite auth database

bendeutsch opened the pull request:
This PR replaces the plain text `auth.txt` file for the builtin auth handler with an SQLite database.

Design goals:
* change as little as possible from the `auth.lua` code, only replace the loading & saving of the text file
* maintain database compatibility with the [sauth](https://github.com/shivajiva101/sauth) mod (**Update**: compatibility has been deliberately broken in favor of sounder database design)
* otherwise treat the auth database the same as the other databases

Since the sauth schema does not have a unique constraint on the player name, but instead uses an autogenerated numerical primary key, I've kept the same schema here. I've added the `id` field to the `authEntry` struct which will mirror the `auth_table` entries in lua, and use the `id` to identify the DB entry to update (deletion is still done directly by name, though).

Attempts to address #6784

bendeutsch commented:
Going is harder than I thought – this is my first time working with SQL in a statically typed language :stuck_out_tongue: At the moment it basically compiles, it doesn't do anything yet.

bendeutsch commented:
@nerzhul yes, I'll probably try to follow the other (two?) databases on this one, though this would mean that I'll need an `AuthDatabase` implementation for the "files" backend as well. Maybe pg too, though this should probably be a separate PR.

In fact, that's a good idea: I should rewrite the current pure-Lua part with such a wrapper. Then I can test the glue code better and can better converge on a common set of functionality.

SmallJoker commented:
Whereas I love the attempt on implementing SQLite into our auth system, I would prefer more generic API functions which would make SQLite databases safely available for easy use in mods. I already considered bundling [lsqlite3](http://lua.sqlite.org/index.cgi/doc/tip/doc/lsqlite3.wiki#overview) (MIT license), but this would require various changes to make it safe for use. The most important operations are basically opening the database with path checks, executing a statement, reading all rows and getting the error code/message.
This allows a separation of the Auth (Lua) and SQLite (C++) code and makes it possible to add the required "glue", like the database migration entirely to Lua.

nerzhul commented:
@SmallJoker for me it's not good to have a such sensitive thing in lua. The best thing we can have is shared auth between multiple servers, and PostgreSQL is the best candidate

bendeutsch commented:
@SmallJoker after reading your ideas, I was about to suggest a more abstract minimal ORM for Lua, but actually I like the idea of giving mods a generic SQL database for their own use, and I can't think of a downside yet.

But not for auth; I was actually amazed when I first learned that any part of the user+password stuff was in Lua. But I guess it makes sense for mod-defined alternatives.

@nerzhul shared auth between multiple servers still has the problem that the permissions are fully stored in the database. This is fine if you want someone who has agreed to the user agreement to have `interact` on all servers, but if the same community wants to set up a survival and a creative server, players will either have `free_move` on both or on neither. But that's a problem for another time, I guess.

nerzhul commented:
@bendeutsch i think it can be time to move those perms to player db

raymoo commented:
Can you even access a sqlite db over the network? Or would all the servers have be running on the same machine?

nerzhul commented:
sqlite is not a network db, it's why pg is important here. also sqlite is not for concurrent binaries as the DB state is in the caller memory

bendeutsch commented:
Latest push includes a `AuthDatabaseFiles` version which mimics the original `auth.lua` in that it reads the `auth.txt` file on start and writes the entire file for each save. It appears to work when reading and rewriting the file.

Further testing will need the Lua-to-C++ glue layer(s); working on that.

bendeutsch commented:
Here's the first part of the Lua-to-C++ glue layer. The auth database is accessed via the `ServerEnvironment` and `ModApiServer` (in `src/script/lua_api/l_server.cpp`). Loading and saving the `auth.txt` is deactivated in `auth.lua`, and only the `get_auth` method connects to the new C++ classes, so most of the auth functionality is broken in this push. Don't use this in production.

My main reason for pushing this small upgrade is to get feedback on whether or not I'm adding this in the correct location, specifically I don't seem to need anything in `src/script/cpp_api` and I'm wondering if I missed something. So, feedback welcome!

Dumbeldor commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,107 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+	return writeAuthFile();
+}
+
+bool AuthDatabaseFiles::deleteAuth(std::string &name)
```
const std::string &name

bendeutsch replied:
Fixed, thanks!

Dumbeldor commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,132 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	return true;
+}
+
+bool AuthDatabaseSQLite3::deleteAuth(std::string &name)
```
const

bendeutsch replied:
Fixed.

bendeutsch commented:
I think the `files` version is finished now. Implementing this has given me some more insights into the interface, so although the SQLite version looks pretty complete, it isn't. I haven't even run it yet :-P

Reading the auth file, creating entries, setting and revoking privileges and deleting players seems to work. Changing the password seems to *not* work, since it creates what look like old hashes, but then again the behaviour is the same in `master`, so that's no fault of my code it seems.

Some thoughts:
- I've kept the `auth_table` in `auth.lua`, even though it technically is not needed anymore, but one change at a time.
- There is some stuff in `get_auth` which looks like it belongs in `create_auth`?
- I'm missing (or overlooking) some basic utility functions. We have two string splitting methods, but no join? Also I'd like a `getstringsetfield` Lua utility function to go with `getstringlistfield`. I may have to add those.
- I'm using sets for the privileges in C++, where I should probably just use vectors for much greater convenience. On the other hand I'm using lists in Lua (tables with numerical keys) where I should be using sets (tables with `true` values for each key). These may change at some point.
- Where are `reload` and `iterate` used?

bendeutsch commented:
The SQLite backend seems to be working. A word of warning if you want to try it out: at the moment the auth backend is hardwired to `sqlite3`, and there's no migration, so it will create an empty `auth.sqlite` file on start.

There should be at least a setting for the auth backend coming, and I may be able to crib a migration from the other DB examples.

bendeutsch commented:
The setting (`auth_backend` in `world.mt`) is working. There's a new command line argument `--migrate-uth` that mirrors the other `migrate` functionalities. This works in my small test case (files to SQLite3), with some caveats:

- There is no backup or deletion of the old DB. The `auth.txt` file remains in place (but will be ignored due to `world.mt`).
- The new DB is not cleared, it is assumed to be empty at the beginning of the migration. If it isn't, there will be duplicate entries or other similar errors.
- Migrating to `files` will probably take `O(n²)` time since the entire `auth.txt` is saved in its entirety after each migrated entry.

However, the migration should be moderately safe due to the flipped setting in `world.mt`, making double migrations abort cleanly, and due to the fact that this is only done once, by hand.

I consider this ready to be pulled, and have removed the WIP from the title. If someone could remove the WIP tag I would be grateful.

SmallJoker commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+			      "`password` VARCHAR(512),"
+			      "`privileges` VARCHAR(512),"
+			      "last_login INTEGER);",
+			      NULL, NULL, NULL),
```
Wrong formatting. Ignore Lint if they requested this.

SmallJoker replied:
I guess you'll have to edit it manually until clang adds more options so we can correct the issues with the code formatting rules.

nerzhul replied:
if needed add // clang-format off before the function and // clang-format on after

SmallJoker commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	PREPARE_STATEMENT(write, "UPDATE auth set name = ?, password = ?, privileges = "
+				 "?, last_login = ? WHERE id = ?;");
+	PREPARE_STATEMENT(create, "INSERT INTO auth (name, password, privileges, "
+				  "last_login) VALUES (?, ?, ?, ?);");
```
Wrong formatting here and two lines above too.

SmallJoker commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	int64_to_sqlite(m_stmt_create, 4, authEntry.last_login);
+	sqlite3_vrfy(sqlite3_step(m_stmt_create), SQLITE_DONE);
+	sqlite3_reset(m_stmt_create);
+	// write back to original authEntry
```
Are you getting this auth again only to get the new `id`?

bendeutsch replied:
I think so, yes. And technically to get any other changes of the DB entry, but in reality we know that there won't be any, true. I'll see what I can do.

SmallJoker commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+void AuthDatabaseSQLite3::initStatements()
+{
+	PREPARE_STATEMENT(read, "SELECT id, name, password, privileges, last_login FROM "
+				"auth WHERE name = ?");
```
SELECT and UPDATE: add `LIMIT 1`. It doesn't have to check the entire database.

Only update: please use `SET` in capitals (trivial)

nerzhul replied:
NO LIMIT 1 should NEVER be used to when you have a primary this is a SQL non sense

nerzhul replied:
@bendeutsch go ahead for a UNIQUE INDEX on the table.
sauth compat is not required here, it's a core backend. Also we can import data outside sauth and import it without problems (except dup names which is a problem in all cases)

SmallJoker commented on src/script/lua_api/l_server.cpp:
```diff
@@ -505,6 +506,191 @@ int ModApiServer::l_set_last_run_mod(lua_State *L)
+	}
+	AuthDatabase *auth_db = server_environment->getAuthDatabase();
+	if (!auth_db) {
+		lua_pushnil(L); // no database (yet)
```
This is very inconsistent. On failure you're pushing nil to Lua, but not on success. I recommend to only return the pointers and push nil in the parent function.

bendeutsch replied:
Yeah, it's very unclean; an attempt to share as much code as possible. I'll change this.

bendeutsch replied:
Changed, also including the `return 0;` from your next review comment.

SmallJoker commented on src/script/lua_api/l_server.cpp:
```diff
@@ -505,6 +506,191 @@ int ModApiServer::l_set_last_run_mod(lua_State *L)
+
+	if (!success) {
+		lua_pushnil(L);
+		return 1;
```
Superfluous. `return 0` does the same, also bracket removal:
```C++
if (!success)
	return 0;
```

bendeutsch replied:
Oh, does this work for all such API functions? I can use this in other places, too.

bendeutsch commented:
**Changed to WIP again**: I just ran into a wierd bug I haven't been able to fix yet: When I change my own privileges, the change does not show up until the *next* privilege change. This works in master, so definitely a blocking change in my code. Stay tuned.

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	SQLOK(sqlite3_exec(m_database,
+			      "CREATE TABLE IF NOT EXISTS `auth` ("
+			      "`id` INTEGER PRIMARY KEY AUTOINCREMENT,"
+			      "`name` VARCHAR(32),"
```
please add an UNIQUE index on the name field as it's unique, this will prvent the ugly limit 1 suggestion

raymoo replied:
Shouldn't it be possible to migrate still?

bendeutsch replied:
Not completely. The code is currently a drop-in for the `sauth` mod, in that it can use the same database file which already contains all the players. But in that case, the `CREATE` statement will not get run, and the DB will continue to not have the unique key in its table definition for all eternity.

nerzhul replied:
having compliant data with sauth is okay.
Scheme: no, please do the right scheme, it's the core engine, we shouldn't be 100% compliant with mod with some design problems

nerzhul replied:
ID is nice, it's better than my player implementation which is lazy and use a string as a primary which is bad in terms of performance for huge db (MT players & auth will not be)

nerzhul replied:
Please also add a dbversion table like in map db permitting to track changes and upgrade the scheme

bendeutsch replied:
I'm sorry, I can't find what you mean with a dbversion table. Can you give a link in GitHub somewhere?

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	// TODO: a better 'join', perhaps?
+	std::ostringstream privstream;
+	privstream << *(authEntry.privileges.begin());
+	for (auto p = std::next(authEntry.privileges.begin());
```
with the std::next ? for (auto &p : authEntry.privileges) is sufficient

nerzhul replied:
in fact i understand and i know what's the problem. Add a table to store the privileges. This string is a non sense in a SGBD.

Table scheme is:

user_privileges (id, privname, primary key(id, privname))

bendeutsch replied:
Ok, will take a while though. I'll write some tests first, this is getting bigger than I can check by hand :-P

nerzhul commented on src/script/lua_api/l_server.cpp:
```diff
@@ -505,6 +506,183 @@ int ModApiServer::l_set_last_run_mod(lua_State *L)
+	if (auth_db->createAuth(authEntry)) {
+		pushAuthEntry(L, authEntry);
+	} else {
+		lua_pushnil(L);
```
return 0 should be sufficient here

bendeutsch replied:
Fixed.

nerzhul commented on src/script/lua_api/l_server.cpp:
```diff
@@ -539,4 +717,11 @@ void ModApiServer::Initialize(lua_State *L, int top)
 	API_FCT(get_last_run_mod);
 	API_FCT(set_last_run_mod);
+
+	API_FCT(auth_read);
```
Can you make auth be a specific Lua object instead of a static function API ? It can be better to have an auth object to manipulate

nerzhul replied:
define a new C++ object, reference it with a LuaRef on the Server stack at initialization like it's done for CSM with "ui" namespace

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -2199,3 +2211,78 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+
+		std::vector<std::string> names_list;
+		srcdb->listNames(names_list);
+		for (std::string &name : names_list) {
```
const std::string &

bendeutsch replied:
Fixed.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -2199,3 +2211,78 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+	}
+
+	try {
+		AuthDatabase *srcdb = ServerEnvironment::openAuthDatabase(
```
can you store AuthDatabase pointers in a unique_ptr please ?
There are some possible leaks here when exceptions are triggered

bendeutsch replied:
Good point. Do you mean only in the `migrate` function, or in general?

nerzhul replied:
At least here as it's a single usage, but it can be used in the server part too i think

bendeutsch replied:
Fixed this at least. Will have to see about the other parts.

nerzhul commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,113 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+		const AuthEntry &authEntry = auth_i.second;
+		output << authEntry.name << ":" << authEntry.password << ":";
+		output << *(authEntry.privileges.begin());
+		for (auto p = std::next(authEntry.privileges.begin());
```
for (const auto &p: authEntry.privileges)

SmallJoker replied:
He implements his own "vector concatenation" function here, which doesn't lead to a superfluous beginning/tailing comma. Because that one would result in an empty string in `str_split`, which is not desired as a priv value.

nerzhul commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,113 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+		std::vector<std::string> privileges = str_split(parts[2], ',');
+		time_t last_login = atol(parts[3].c_str());
+
+		AuthEntry authEntry = {
```
no need for a temp object here, affect it directly

bendeutsch replied:
Done.

SmallJoker commented on src/script/lua_api/l_server.cpp:
```diff
@@ -505,6 +506,183 @@ int ModApiServer::l_set_last_run_mod(lua_State *L)
+	if (!auth_db)
+		return 0;
+	auth_db->reload();
+	return 1;
```
Nothing is returned here.

bendeutsch replied:
Fixed.

SmallJoker commented on src/script/lua_api/l_server.cpp:
```diff
@@ -505,6 +506,183 @@ int ModApiServer::l_set_last_run_mod(lua_State *L)
+		return nullptr;
+	AuthDatabase *auth_db = server_environment->getAuthDatabase();
+	if (!auth_db)
+		return nullptr;
```
Superfluous check and return.

bendeutsch replied:
Condensed.

SmallJoker commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+void AuthDatabaseSQLite3::reload()
+{
+	// noop for SQLite
+	return;
```
Superfluous return

bendeutsch replied:
Fixed, I think I got them all.

SmallJoker commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,113 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+		std::string line;
+		std::getline(file, line);
+		std::vector<std::string> parts = str_split(line, ':');
+		if (parts.size() < 4) // also: empty line at end
```
Just to have it said: Last login exists since almost four years but our auth mechanism didn't add dummy values if they didn't exist.
However, I think we can consider this as a feature to clean up really old, unused accounts.

bendeutsch replied:
I don't understand your comment, sorry. Do you mean that there are `auth.txt` files without the fourth field? I think the 4 was just meant as some kind of sanity check to guard against empty lines. Would it help if I changed it to 3? But what would I use as `last_login` then?

sfan5 replied:
IMO it's not okay for Minetest to quietly drop data, even if it's okay and even useful in most cases

SmallJoker replied:
These players are however still in the player database, so when their auth is deleted, someone could take over this used account. Result: Joining at any unknown place with lack of initial items or full-equipped armor.
@bendeutsch Could you please change this to `parts.size() < 3` and set `last_login` to `0` when it's not available?

nerzhul replied:
yes it's a good idea @SmallJoker

bendeutsch replied:
Done. Sorry for the delay :-(

nerzhul commented:
A very nice nice in the good way

bendeutsch commented:
While we're discussing the database scheme, another idea: since uniqueness of a name is determined *on lower cased versions*, how about adding a new unique column with the lower cased name? I thought about suggesting it for a later PR, but might as well consider it now.

bendeutsch commented:
Newest push has a UNIQUE index on names, unit tests for files and SQLite, and privileges in a separate table. I think the only requested change I still need to look at is the "object instead of functions" request from @nerzhul . This will probably take some time to wrap my head around.

Oh, and I'd need to retest the migration; nothing should have changed, but it isn't covered by unit tests. Or maybe I should do that.

bendeutsch commented:
Last push moves the C++ code of the Lua API to its own files (`lua_api/l_auth.{cpp,h}`) and the Lua functions themselves from (say) `core.auth_read` to `core.auth.read`.

@nerzhul is this what you had in mind, or rather `core.auth:read`? The code currently doesn't need a `self` parameter, but it wouldn't hurt.

bendeutsch commented:
I think I've changed what I can. Removing the "WIP: " title mark, could someone remove the "WIP" label please?

nerzhul commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,106 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+		std::vector<std::string> parts = str_split(line, ':');
+		if (parts.size() < 4) // also: empty line at end
+			continue;
+		std::string &name = parts[0];
```
name & pass should be const

bendeutsch replied:
Done.

nerzhul commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,106 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+		std::string &name = parts[0];
+		std::string &password = parts[1];
+		std::vector<std::string> privileges = str_split(parts[2], ',');
+		time_t last_login = atol(parts[3].c_str());
```
std::time_t

bendeutsch replied:
Done.

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,169 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+bool AuthDatabaseSQLite3::saveAuth(const AuthEntry &authEntry)
+{
+	verifyDatabase();
+
```
you must use a begin and commit function to ensure data integrity there in a single write transaction

bendeutsch replied:
Done here and in `createAuth`, too.

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,169 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	int64_to_sqlite(m_stmt_delete_privs, 1, authEntry.id);
+	sqlite3_vrfy(sqlite3_step(m_stmt_delete_privs), SQLITE_DONE);
+	sqlite3_reset(m_stmt_delete_privs);
+	for(const std::string &privilege : authEntry.privileges) {
```
code style

bendeutsch replied:
What's the exact problem? `clang-format` does not seem to be helping, here…

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,169 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+{
+	verifyDatabase();
+
+	// id autoincrements
```
exactly same problems there.

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,169 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	sqlite3_reset(m_stmt_last_insert_rowid);
+
+	// assume no privileges exist yet for freshly created entry
+	for(const std::string &privilege : authEntry.privileges) {
```
don't duplicate this code part, use a private function to perform the loop accross the two write functions

bendeutsch replied:
Done, private method `writePrivileges`.

nerzhul commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,169 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	int changes = sqlite3_changes(m_database);
+	sqlite3_reset(m_stmt_delete);
+
+	// privileges deleted by foreign key on delete cascade
```
nice :)

sfan5 commented on src/database/database-files.h:
```diff
@@ -41,3 +41,23 @@ class PlayerDatabaseFiles : public PlayerDatabase
+	AuthDatabaseFiles(const std::string &savedir);
+	virtual ~AuthDatabaseFiles() = default;
+
+	virtual bool getAuth(const std::string &name, AuthEntry &res);
```
this can be a const method

bendeutsch replied:
I tried, and it's tricky. Any call will also call `verifyDatabase`, which will potentially create the database, flag everything as `m_initialized` etc.

Yes, I could mark any such things (the statements, too, are changed during reads until a `reset` is called) as `volatile`, but then I'd have to back-patch this into the map and player databases too; none of which, by the way, have `const` methods. Is this really that critical?

sfan5 replied:
It's not critical at all, just something I noticed.

sfan5 commented on src/database/database-files.h:
```diff
@@ -41,3 +41,23 @@ class PlayerDatabaseFiles : public PlayerDatabase
+	virtual bool saveAuth(const AuthEntry &authEntry);
+	virtual bool createAuth(AuthEntry &authEntry);
+	virtual bool deleteAuth(const std::string &name);
+	virtual void listNames(std::vector<std::string> &res);
```
same

sfan5 commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,170 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	PREPARE_STATEMENT(list_names, "SELECT name FROM auth ORDER BY name DESC");
+
+	PREPARE_STATEMENT(read_privs, "SELECT privilege FROM user_privileges WHERE id = ?");
+	PREPARE_STATEMENT(write_privs, "INSERT OR IGNORE INTO user_privileges (id, privilege) VALUES (?, ?)");
```
wouldn't `REPLACE INTO` have the same effect?

nerzhul replied:
a NOOP is better in this case than rewriting the row as there is no other column to alter than the two in the query

sfan5 commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,170 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	res.id = sqlite_to_uint(m_stmt_read, 0);
+	res.name = sqlite_to_string(m_stmt_read, 1);
+	res.password = sqlite_to_string(m_stmt_read, 2);
+	res.last_login = (time_t)sqlite_to_uint(m_stmt_read, 3);
```
is sqlite's uint big enough to hold timestamps beyond 2038?

bendeutsch replied:
Yes: http://www.sqlite.org/datatype3.html

An `INTEGER` is in practice a 64bit type. The Y2k38 problem only affects 32bit integers.

NewbProgrammer101 replied:
Thus, people with 32-bit binaries can't use this functionality beyond 2038? Or something else worse happens...

bendeutsch replied:
No, 32-bit binaries can still use 64bit integers ("longs"), that's not the problem. 

The *actual* problem, which I just found out about (thanks @NewbProgrammer101 for making me check!), is that Minetest's sqlite utility functions *use 32bit integers!!* I'll have to write 64bit variants of these and use them instead. Oh, and `time_t`, as I found out with even more digging, may be 32 bits on some 32-bit systems. So I'll have to change that, too, to be safe.

bendeutsch replied:
Added `int64` methods; `last_login` is now a s64.

raymoo replied:
Just want to note that "long" is 32 bits in 32 bit binaries (usually), it's long long that's guaranteed 64 bits.

nerzhul replied:
s64 or u64 should be used, as they are our internals for that. uint64_t & int64_t are C standard (cstdint) for that

sfan5 commented on src/database/database.h:
```diff
@@ -61,3 +63,25 @@ class PlayerDatabase
+
+struct AuthEntry
+{
+	u64 id;
```
this is just an implementation detail of the sqlite db, but I guess it can't be avoided to have it in here.

bendeutsch replied:
It probably could, now that names are `UNIQUE`, but I've been asked to keep a numerical `id` anyway.

It may make a difference for future PostgreSQL implementations.

sfan5 commented on src/script/lua_api/l_auth.cpp:
```diff
@@ -0,0 +1,216 @@
+	// privileges
+	lua_newtable(L);
+	int privtable = lua_gettop(L);
+	for (const std::string &privs : authEntry.privileges) {
```
can be `auto`

bendeutsch replied:
I would argue that in this case, writing out the actual type aids clarity (the '.c_str' method is used two lines below).

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -413,6 +413,18 @@ ServerEnvironment::ServerEnvironment(ServerMap *map,
+
+	std::string auth_name = "files";
+	if (conf.exists("auth_backend")) {
+		conf.getNoEx("auth_backend", auth_name);
```
does this even work without a `&`?

bendeutsch replied:
The `getNoEx` method is defined as
```bool getNoEx(const std::string &name, std::string &val) const;
```
The second parameter is pass-by-reference, and not const, so it can be assigned to.

It also works for the player backend a few lines above, which already existed in code.

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -2199,3 +2211,75 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+
+		std::vector<std::string> names_list;
+		srcdb->listNames(names_list);
+		for (const std::string &name : names_list) {
```
`const auto`

bendeutsch replied:
Again, as above, I would keep it to help future readers better understand the code.

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -2199,3 +2211,75 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+			success = success && dstdb->createAuth(authEntry);
+			if (!success) {
+				actionstream << "Failed to migrate " << name << std::endl;
+			}
```
brace removal

bendeutsch replied:
Done.

bendeutsch commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,106 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+{
+	res.clear();
+	res.reserve(m_auth_list.size());
+	for (auto const &res_pair : m_auth_list) {
```
Done.

bendeutsch commented on src/database/database-sqlite3.cpp:
```diff
@@ -606,3 +606,170 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+	int64_to_sqlite(m_stmt_delete_privs, 1, authEntry.id);
+	sqlite3_vrfy(sqlite3_step(m_stmt_delete_privs), SQLITE_DONE);
+	sqlite3_reset(m_stmt_delete_privs);
+	for(const std::string &privilege : authEntry.privileges) {
```
Fixed.

bendeutsch commented:
Fixed, rebased, and edited the description (it still promised compatibility with sauth).

To my knowledge, all open points have been addressed. If anyone has a comment, request or correction that I did not acknowledge: please let me know, this was not on purpose.

SmallJoker commented on src/script/lua_api/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 set(common_SCRIPT_LUA_API_SRCS
 	${CMAKE_CURRENT_SOURCE_DIR}/l_areastore.cpp
+	${CMAKE_CURRENT_SOURCE_DIR}/l_auth.cpp
```
Also add to https://github.com/minetest/minetest/blob/master/build/android/jni/Android.mk

bendeutsch replied:
Done, but I have not been able to successfully build on Android yet, so I can't test this.

I see that not all unit test files are listed. Is this on purpose?

SmallJoker replied:
Probably it compiled fine on Android so nobody noticed that they were forgotten - or it's on purpose and the documentation is missing. :man_shrugging:

SmallJoker commented on src/script/lua_api/l_auth.cpp:
```diff
@@ -0,0 +1,216 @@
+}
+
+// auth_read(name)
+int ModApiAuth::l_auth_read(lua_State *L)
```
These functions need to be documented.

bendeutsch replied:
In this case, they're exposed via a `core.auth` object, as per a previous request. In addition, they're not supposed to be accessed by mod authors… I mean, I never thought of them as… huh.

Maybe it *does* make sense for mods to access this? Even though it exposes (hashed) passwords? Then documentation is in order. Otherwise, how do I let this be used by `builtin/game/auth.lua`, but nothing else?

bendeutsch replied:
Ok, I'll do that. Where should the documentation of the (then private) object+methods go? Is `l_auth.{h,cpp}` ok?

SmallJoker replied:
@bendeutsch The function arguments are already mentioned there in the comments. When the functions are unavailable to modders, this does not seem to need more documentation.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -2199,3 +2211,74 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+
+	} catch (BaseException &e) {
+		errorstream << "An error occured during migration: " << e.what()
+			    << std::endl;
```
Mixed indents. Please only use tabs

bendeutsch replied:
Done.

NewbProgrammer101 commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,106 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+
+AuthDatabaseFiles::AuthDatabaseFiles(const std::string &savedir) : m_savedir(savedir)
+{
+	readAuthFile();
```
You want the whole thing to be from a constructor that gets called multiple times when an object is made?

Try something sane and have `AuthDatabaseFiles` as a global object and call `readAuthFile()` directly. For example:

( Just an example. don't use this code, it's broken )
```
extern std::unique_ptr<AuthDatabaseFiles> sqlauth; // Declare the global object. Should be in a header that does global stuff. `std::unique_ptr` to prevent memory leaks.

std::unique_ptr<AuthDatabaseFiles> sqlauth = nullptr; // In a .cpp file for the header mentioned above.

// In this file, you can just have something call `readAuthFile()` by doing...
sqlauth = std::unique_ptr<AuthDatabasefiles>( new AuthDatabaseFiles() );
sqlauth->readAuthFile();
```

Boom.

bendeutsch replied:
The `AuthDatabase` is created once per world, when the server environment is created, so I don't think it's unreasonable to read the file in the constructor.

I do seem to be missing a delete, though, so I'll either add that or more likely use a smart pointer in the serverenvironment.

bendeutsch replied:
I added a `delete`, on par with the player database a few lines above.

NewbProgrammer101 commented:
Overall, this looks :ok_hand:

paramat commented:
This seems high priority for 5.0.0 considering all the issues caused by auth.txt.

nerzhul commented:
@paramat yes it's definitively a requirement to release 5.0.0 as we started to migrate the player backend, this is the final work on having consistent player backends.
Just a note: 5.0.0 will require this PR with postgresql support added in a future PR, i will work on it after this implementation is merged. This iwll permits server owners to have mutualized auth db

nerzhul commented:
it's not a bugfix it's a feature

bendeutsch commented:
@paramat I'm guessing it addresses #6783 for both backends and #5334 when using the SQLite backend (the files backend still writes on each change in this version).

paramat commented:
Yes those issues is why i added 'bugfix'.

sfan5 commented on src/util/string.h:
```diff
@@ -704,3 +704,22 @@ inline const std::string duration_to_string(int sec)
+inline std::string str_join(const std::vector<std::string> &list,
+		const std::string &delimiter)
+{
+	std::stringstream sstr;
```
should be an `std::ostringstream`

bendeutsch replied:
Done.

nerzhul commented on src/unittest/test_authdatabase.cpp:
```diff
@@ -0,0 +1,309 @@
+	SQLite3Provider(const std::string &dir) : dir(dir){};
+	virtual ~SQLite3Provider()
+	{
+		if (auth_db)
```
if (x) delete x: conditional is not needed

bendeutsch replied:
Fixed; four occurances.

bendeutsch commented on src/unittest/test_authdatabase.cpp:
```diff
@@ -0,0 +1,309 @@
+	};
+	virtual AuthDatabase *getAuthDatabase()
+	{
+		if (auth_db)
```
Fixed (in the same update above).

bendeutsch commented:
Just rebased and fixed a formatting problem. Is there anything still pending here? @nerzhul , it still says you have requested changes – I think everything has been addressed? Or am I missing something?

nerzhul commented on src/database/database-files.cpp:
```diff
@@ -177,3 +178,106 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
+
+bool AuthDatabaseFiles::deleteAuth(const std::string &name)
+{
+	auto count = m_auth_list.erase(name);
```
you can remove the temp variable here

bendeutsch replied:
Removed and folded into the condition.

nerzhul commented on src/database/database-files.h:
```diff
@@ -41,3 +41,23 @@ class PlayerDatabaseFiles : public PlayerDatabase
+	virtual void reload();
+
+private:
+	std::map<std::string, AuthEntry> m_auth_list;
```
no need to order names, you can use std::unordered_map

bendeutsch replied:
Good point, changed.

nerzhul commented on src/script/lua_api/l_auth.cpp:
```diff
@@ -0,0 +1,216 @@
+	if (auth_db->createAuth(authEntry)) {
+		pushAuthEntry(L, authEntry);
+		return 1;
+	} else {
```
you can remove the else and just do a return 0

bendeutsch replied:
Changed.

bendeutsch commented:
1. Will do.
2. We should probably assume `files` if the setting is not present, but `auth.txt` is? Will take a bit of work, though.
3. Normalization, and potential preparations for moving the privileges from auth to the player DB, where they probably make more sense. Or split into account / world level privileges. But that's another story.
4. After migration, sure.

bendeutsch commented:
@SmallJoker :
1. Documentation has been added, please check.
2. I now understand you were talking about migration (regular use already falls back to "files"). I've changed the migration to use "files" as a fallback, with a warning.
3. (this line intentionally left blank for number reasons)
4. I've thought about this, and technically this needs to work for many backends (you can migrate the other way, too). Backends should have a `deactivateStorage` method, or something, and one would need to check that the "backup name" is not already taken, or auto-increment, or something.

So `auth.txt.bak` is not quite so simple, except if we reduce it to "only if migrating from files, and only if there is no `auth.txt.bak` already". Would that work?

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -2274,3 +2287,73 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
+			success = srcdb->getAuth(name, authEntry);
+			success = success && dstdb->createAuth(authEntry);
+			if (!success)
+				actionstream << "Failed to migrate " << name << std::endl;
```
Should be at least a warning

SmallJoker commented:
The documentation looks good now, thanks.
> So auth.txt.bak is not quite so simple, except if we reduce it to "only if migrating from files, and only if there is no auth.txt.bak already". Would that work?

Yes, that sounds good.
PS: The commits can be squashed on merge automatically - they would help a bit to keep track of the requested changes and what was done to satisfy them.

bendeutsch commented:
New commit now (safely) moves `auth.txt` to `auth.txt.bak`

I know about the squashing thing, but I have also had PRs exceed 50 commits and run for months. Rebasing becomes a complete nightmare if you have to figure out how your old (and since replaced) code would have adapted to the new changes.

I think I'll start using discreet commits again, but rolling up all but the last three or so. Let's see how this works.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -2350,9 +2350,28 @@ bool ServerEnvironment::migrateAuthDatabase(
 	} catch (BaseException &e) {
 		errorstream << "An error occured during migration: " << e.what()
-				<< std::endl;
+			    << std::endl;
```
Mixed indents again

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -2350,9 +2350,28 @@ bool ServerEnvironment::migrateAuthDatabase(
+			else
+				warningstream << "auth.txt.bak already exists, auth.txt "
+						 "not renamed" << std::endl;
+		} else errorstream << "NOT FILES?" << std::endl;
```
Probably for debug purposes. Should be removed because another backend might come soon.

SmallJoker commented:
@bendeutsch Usually it takes a few weeks until the PR has to be rebased (depending on the touched code). If the change requests were already handled some time ago, then it's of course fine to squash before rebasing so that conflict resolving gets easier.

rubenwardy commented:
🎉
